nano/node/node.cpp

```diff
@@ -3477,9 +3519,94 @@ void nano::active_transactions::update_difficulty (nano::block const & block_a)
+		{
+			roots.modify (existing, [difficulty](nano::conflict_info & info_a) {
+				info_a.difficulty = difficulty;
+			});
+			adjust_difficulty (block_a.hash ());
+		}
+	}
+}
+
+void nano::active_transactions::adjust_difficulty (nano::block_hash const & hash_a)
+{
+	assert (!mutex.try_lock ());
+	std::deque<std::pair<nano::block_hash, int64_t>> remaining_blocks;
+	remaining_blocks.push_back (std::make_pair (hash_a, 0));
+	std::unordered_set<nano::block_hash> processed_blocks;
+	std::vector<std::pair<nano::uint512_union, int64_t>> elections_list;
+	uint128_t sum (0);
+	while (!remaining_blocks.empty ())
+	{
+		auto item (remaining_blocks.front ());
+		remaining_blocks.pop_front ();
+		auto hash (item.first);
+		// item.second is level
+		if (processed_blocks.find (hash) == processed_blocks.end ())
+		{
+			auto existing (blocks.find (hash));
+			if (existing != blocks.end () && !existing->second->confirmed && !existing->second->stopped && existing->second->status.winner->hash () == hash)
+			{
+				auto previous (existing->second->status.winner->previous ());
+				if (!previous.is_zero ())
+				{
+					remaining_blocks.push_back (std::make_pair (previous, item.second + 1));
+				}
+				auto source (existing->second->status.winner->source ());
+				if (!source.is_zero () && source != previous)
+				{
+					remaining_blocks.push_back (std::make_pair (source, item.second + 1));
+				}
+				auto link (existing->second->status.winner->link ());
+				if (!link.is_zero () && !node.ledger.is_epoch_link (link) && link != previous)
+				{
+					remaining_blocks.push_back (std::make_pair (link, item.second + 1));
+				}
+				for (auto & dependent_block : existing->second->dependent_blocks)
+				{
+					remaining_blocks.push_back (std::make_pair (dependent_block, item.second - 1));
+				}
+				processed_blocks.insert (hash);
+				nano::uint512_union root (previous, existing->second->status.winner->root ());
+				auto existing_root (roots.find (root));
+				if (existing_root != roots.end ())
+				{
+					sum += existing_root->difficulty;
+					elections_list.push_back (std::make_pair (root, item.second));
+				}
+			}
+		}
+	}
+	if (elections_list.size () > 1)
+	{
+		uint64_t average (static_cast<uint64_t> (sum / elections_list.size ()));
+		// Potential overflow check
+		uint64_t divider (1);
+		if (elections_list.size () > 1000000 && (average - node.network_params.publish_threshold) > elections_list.size ())
+		{
+			divider = ((average - node.network_params.publish_threshold) / elections_list.size ()) + 1;
+		}
+		// Set adjusted difficulty
+		for (auto & item : elections_list)
+		{
+			auto existing_root (roots.find (item.first));
+			uint64_t difficulty_a (average + (item.second / divider));
+			roots.modify (existing_root, [difficulty_a](nano::conflict_info & info_a) {
+				info_a.adjusted_difficulty = difficulty_a;
+			});
+		}
+	}
+	// Return difficulty instead of adjusted value
+	else if (elections_list.size () == 1)
```

Comment:
Is `Return` the correct word to use here?

Reply:
Thanks